pkg/build/generator/generator.go

```diff
@@ -258,24 +288,57 @@ func (g *BuildGenerator) resolveImageRepoReference(ctx kapi.Context, from *kapi.
 	} else {
 		namespace = defaultNamespace
 	}
-	imageStream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), from.Name)
-	if err != nil {
-		if errors.IsNotFound(err) {
+
+	switch from.Kind {
+	case "ImageStreamImage":
+		bits := strings.Split(from.Name, "@")
+		stream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), bits[0])
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", fatalError{err}
+		}
+		ref, err := imageapi.DockerImageReferenceForStream(stream)
+		if err != nil {
 			return "", err
 		}
-		return "", fatalError{err}
-	}
-
-	if imageStream == nil || len(imageStream.Status.DockerImageRepository) == 0 {
-		return "", fmt.Errorf("could not resolve image stream %s/%s", namespace, from.Name)
-	}
-	glog.V(4).Infof("Found image stream %s/%s", namespace, imageStream.Name)
+		ref.ID = bits[1]
+		return ref.String(), nil
+	case "ImageStreamTag":
+		bits := strings.Split(from.Name, ":")
+		stream, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), bits[0])
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return "", err
+			}
+			return "", fatalError{err}
```

Comment:
this logic is tightly coupled to the build controller, so there really isn't an "above".  Returning this as fatal tells the controller "if you retry this, it's just going to fail again, so don't bother" which is easier than forcing the controller to understand all the reason codes that might come back so it can decide if it's fatal.  We're making the decision as early as possible, as close to the reason for the failure as possible.